contracts/exchange/contracts/src/MixinExchangeCore.sol

```diff
@@ -448,5 +562,27 @@ contract MixinExchangeCore is
             order.feeRecipientAddress,
             fillResults.makerFeePaid
         );
+
+        // Transfer protocol fee -> staking if the fee should be paid
+        if (staking != address(0)) {
+            // If sufficient ether was sent to the contract, the protocol fee should be paid in ETH.
+            // Otherwise the fee should be paid in WETH.
+            uint256 protocolFee = fillResults.protocolFeePaid;
+            if (address(this).balance >= protocolFee) {
+                IStaking(staking).payProtocolFee.value(protocolFee)(order.makerAddress);
+            } else {
+                // Transfer the Weth
+                _dispatchTransferFrom(
+                    orderHash,
+                    WETH_ASSET_DATA,
+                    takerAddress,
+                    staking,
+                    protocolFee
+                );
+
+                // Attribute the protocol fee to the maker
+                IStaking(staking).recordProtocolFee(order.makerAddress, protocolFee);
```

Comment:
Thinking about it more, I kind of like having the `weth.transferFrom` call happen within `staking.recordProtocolFee` rather than the `Exchange` contract. This change drastically reduces the amount of trust that the `Staking` contract must have in the `Exchange` contract. It only slightly increases the amount of trust the `ERC20Proxy` has in the `Staking` contract, since it already needs to be authorized to transfer ZRX.